ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/FunctionArgumentsUtil.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.util;
+
+import io.confluent.ksql.execution.codegen.TypeContext;
+import io.confluent.ksql.execution.expression.tree.Expression;
+import io.confluent.ksql.execution.expression.tree.FunctionCall;
+import io.confluent.ksql.execution.expression.tree.LambdaFunctionCall;
+import io.confluent.ksql.function.GenericsUtil;
+import io.confluent.ksql.function.KsqlScalarFunction;
+import io.confluent.ksql.function.UdfFactory;
+import io.confluent.ksql.function.types.GenericType;
+import io.confluent.ksql.function.types.LambdaType;
+import io.confluent.ksql.function.types.ParamType;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlLambda;
+import io.confluent.ksql.schema.ksql.types.SqlLambdaResolved;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.Pair;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public final class FunctionArgumentsUtil {
+
+  private FunctionArgumentsUtil() {
+
+  }
+
+  public static class FunctionArgumentsAndContext {
+    private final List<TypeContext> contexts;
```

Comment:
Couple things about `TypeContext`:

1. I don't think you need `visitType` in there anymore. Seems unused.
2. I think you could simplify `TypeContext` by just building the mapping from lambda variable to type in `getLambdaContextAndType` and creating a child `TypeContext` with this mapping, e.g.:

```
class TypeContext {
    // no list of types
    final Map<String, SqlType> lambdaInputTypeMapping;

    public TypeContext copyWithLambdaVariableTypeMapping(final Map<String, SqlType> variableToType) {
        final HashMap<String, SqlType> mapping = new HashMap<>(lambdaInputTypeMapping);
        for (final String v : variableToType.keySet()) {
          if (mapping.containsKey(v) && !mapping.get(v).equals(variableToType.get(v))) {
              throw new IllegalStateException();
          }
          mapping.put(v, variableToType.get(v));
        }
        return new TypeContext(ImmutableMap.copyOf(mapping));
    }
}

Then, in `getLambdaContextAndType` you can do:
```
if (expression instanceof LambdaFunctionCall) {
          final LambdaFunctionCall lambdaFunctionCall = (LambdaFunctionCall) expression;

          // the function returned from the UDF factory should have lambda
          // at this index in the function arguments if there's a
          // lambda node at this index in the function node argument list
          if (!(parameter instanceof LambdaType)) {
            throw new RuntimeException("Error while processing lambda function. "
                + "This is most likely an internal error and a "
                + "Github issue should be filed for debugging. "
                + "Include the function name, the parameters passed in, the expected "
                + "signature, and any other relevant information.");
          }

          final LambdaType lambdaParameter = (LambdaType) parameter;
          final Iterator<String> variablesI = lambdaFunctionCall.getArguments().listIterator();
          final HashMap<String, SqlType> variableTypeMapping = new HashMap<>();
          for (ParamType inputParam : lambdaParameter.inputTypes()) {
            if (inputParam instanceof GenericType) {
              final GenericType genericParam = (GenericType) inputParam;
              if (!reservedGenerics.containsKey(genericParam)) {
                throw new RuntimeException(
                    String.format(
                        "Could not resolve type for generic %s. "
                            + "The generic mapping so far: %s",
                        genericParam.toString(), reservedGenerics.toString()));
              }
              variableTypeMapping.put(variablesI.next(), reservedGenerics.get(genericParam));
            } else {
              variableTypeMapping.put(
                  variablesI.next(),
                  SchemaConverters.functionToSqlConverter().toSqlType(inputParam)
              );
            }
          }
          final TypeContext childContext = context.copyWithMapping(variableTypeMapping);
          final SqlType resolvedLambdaReturnType =
              expressionTypeManager.getExpressionSqlType(expression, childContext);
```

It's not a big change, but I think it helps readability because the relationship between the lambda parameters in the registered function and those in the expressions is tracked in one place (rather than this intermediate list and map in `TypeContext`.

It also makes `TypeContext` immutable which makes it easier to reason about.